### PR TITLE
feat(txpool) modify txpool guard to be for pipeline syncs only

### LINKED ISSUE #4066: Use different sync controll mechanism in txpool
### Describe the feature

ref https://github.com/paradigmxyz/reth/pull/4063#issuecomment-1665649662

txpool is effectively disabled rn when we're syncing, the motivation was that txpool is redundant while pipeline (or rather initial sync).

this becomes more tricky during live sync, so we should no longer rely on the syncing flag to guard txpool

the simplest solution would be only to use it until the first time it's false

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by 0xprames**
should close #4066 

the tests in `network/src/transactions.rs` need a bit of work - I think there's some bugs hiding in them rn.

### CODE REVIEW

File: crates/net/network/src/transactions.rs [hunk 1]
```diff
@@ -808,10 +808,65 @@ mod tests {
...
+        transactions.on_network_tx_event(NetworkTransactionEvent::IncomingTransactions {
+            peer_id,
+            msg: Transactions(vec![TransactionSigned::default()]),
+        });
+        assert!(!NetworkInfo::is_initially_syncing(&handle));
+        assert!(NetworkInfo::is_syncing(&handle));
+        // assert!(!pool.is_empty()); this keeps failing - indicating that there's a bug with these
+        // tests
+
+        // if we attempt to modify the initial test above with SyncState::Idle, the same assert
```
Comment (0xprames) [inline crates/net/network/src/transactions.rs:857]: cc @mattsse re these tests

tl;dr - `assert pool.is_empty()` and `assert !pool.is_empty()` don't seem to be doing what we want it to in these tests (?)

File: crates/net/network-api/src/lib.rs [hunk 2]
```diff
@@ -49,6 +49,9 @@ pub trait NetworkInfo: Send + Sync {
 
     /// Returns `true` if the network is undergoing sync.
     fn is_syncing(&self) -> bool;
+
+    /// Returns `true` if the network is undergoing an initial sync.
```
Comment (mattsse) [inline crates/net/network-api/src/lib.rs:53]: This needs some additional context what this exactly represents wrt pipeline sync:
if this is a fresh node sync then this is expected to track the first complete pipeline run.

File: crates/net/network/src/network.rs [hunk 3]
```diff
@@ -55,6 +55,7 @@ impl NetworkHandle {
             network_mode,
             bandwidth_meter,
             is_syncing: Arc::new(AtomicBool::new(false)),
+            initial_sync_done: Arc::new(AtomicBool::new(false)),
```
Comment (mattsse) [inline crates/net/network/src/network.rs:58]: we could put both states in a single AtomicU8, but perhaps the added complexity is not worth it since we only ever use relaxed.
I'd be okay with both solutions

File: crates/net/network/src/network.rs [hunk 4]
```diff
@@ -247,18 +248,38 @@ impl NetworkInfo for NetworkHandle {
...
+        let init_sync_done = self.inner.initial_sync_done.load(Ordering::Relaxed);
+        self.inner.is_syncing.store(future_state, Ordering::Relaxed);
+        if !init_sync_done {
+            // we've either been in Idle or pipeline Syncing here, check to see if we're
+            // moving from a pipeline sync to idle
+            if curr_state && !future_state {
+                // set initial_sync_done flag to true here, every sync after this will be a livesync
+                self.inner.initial_sync_done.store(true, Ordering::Relaxed);
+            }
+        }
```
Comment (mattsse) [inline crates/net/network/src/network.rs:282]: I think we can simplify this,
basically `initial_sync_done` will be true when `is_syncing` goes from false to true

I think we can use `.swap` here and compare the replaced value

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, I first envisioned that the `initial_sync_done` flag will be tracked in the transactionmanager, but I think we can also just move it to the Syncstate impl, perhaps this is useful somewhere else.

ATM this seems perfectly reasonable but perhaps the txpool needs additional logic around this, but I can't think of anything else to check and just using the initial sync as the guard is sufficient atm.
so I'm supportive of adding it to syncstate.

tracking another atomic bool is inexpensive so this is fine, but we could put all of them in one atomicu8

File: crates/net/network/src/network.rs (hunk 4 above)
Comment (0xprames) [reply crates/net/network/src/network.rs:282]: mm let me think about this - I think I see what you're saying, but I need think through the implementation

I think you're right in that we know when `is_syncing` goes from false-> true (the first time), `initial_sync_done` -> true

Comment (0xprames) [reply crates/net/network/src/network.rs:282]: yeah I think this will replace the block that I have now:
```
self.inner.initial_sync_done.compare_exchange(false, curr_state && !future_state, Ordering::Relaxed, Ordering::Relaxed);
```

basically a CAS (`compare_and_swap` seems deprecated in favor of this new `compare_exchange` api)  that only sets `initial_sync_done` if its false

I do want to mention that `initial_sync_done` should only be true when `!init_sync_done` && `is_syncing` goes from true->false

what you mentioned may not work (when `is_syncing` goes from false->true) because I think we need to capture when the initial_sync completes and not just when it starts.


otherwise in the network manager code we won't be able to differentiate between a live_sync and an ongoing pipeline sync (i.e is_syncing() && is_initial_syncing()` will be true even during the pipeline sync

Comment (0xprames) [reply crates/net/network/src/network.rs:282]: what we want to capture is the state transition from false->true->false with `is_syncing`

false->true show cases that we've kicked off a pipeline sync, but we can't mark `is_initial_done` until it has actually completed

File: crates/net/network/src/network.rs (hunk 3 above)
Comment (0xprames) [reply crates/net/network/src/network.rs:58]: yeah for sure - I agree in that this may be a bit more naive, but it's cheap for us to use another AtomicBool and we get the added benefit of readability/simplicity

File: crates/net/network-api/src/lib.rs [hunk 5]
```diff
@@ -49,6 +49,9 @@ pub trait NetworkInfo: Send + Sync {
 
     /// Returns `true` if the network is undergoing sync.
     fn is_syncing(&self) -> bool;
+
+    /// Returns `true` when a fresh node is initially being synced, known as a Pipeline sync
```
Comment (mattsse) [inline crates/net/network-api/src/lib.rs:53]: this is not quite accurate because this also returns true on restart if the first pipeline sync is active.

File: crates/net/network/src/network.rs [hunk 6]
```diff
@@ -247,18 +248,38 @@ impl NetworkInfo for NetworkHandle {
...
+        self.inner.is_syncing.store(future_state, Ordering::Relaxed);
+        // explicitly throw this error away since we expect it to error every time after it's been
+        // set to true, i.e after every live sync
+        let _ = self.inner.initial_sync_done.compare_exchange(
+            false,
+            // on first move back from Syncing->Idle - swap and set init_sync_done to true
+            curr_syncing && !future_state,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        );
```
Comment (mattsse) [inline crates/net/network/src/network.rs:282]: this is not quite exactly what I had in mind, you can do a swap on line 272 and then update is_initial_done depending on the values

**Review by mattsse — CHANGES_REQUESTED** [review]
good progress almost there

File: crates/net/network/src/network.rs (hunk 6 above)
Comment (0xprames) [reply crates/net/network/src/network.rs:282]: ah.. ok yeah. I misunderstood the initial comment oops, but Ill keep the CAS bc it looks cleaner (and setting `init_sync_done` it in one atomic step is nice)

basically what you're saying is instead of the load/store on `is_syncing` I just need one swap and the return value is what is being set as `curr_syncing` (which I'm going to rename to `prev_state` to make a bit clearer)

File: crates/net/network-api/src/lib.rs (hunk 5 above)
Comment (0xprames) [reply crates/net/network-api/src/lib.rs:53]: because during a pipeline sync, state of the sync "job" is stored externally I think, and it'll be loaded in/continued from upon restart

File: crates/net/network/src/network.rs [hunk 7]
```diff
@@ -247,18 +248,37 @@ impl NetworkInfo for NetworkHandle {
...
+        let prev_state = self.inner.is_syncing.swap(future_state, Ordering::Relaxed);
+        // explicitly throw this error away since we expect it to error every time after it's been
+        // set to true, i.e after every live sync
+        let _ = self.inner.initial_sync_done.compare_exchange(
+            false,
+            // on first move back from Syncing->Idle - swap and set init_sync_done to true
+            prev_state && !future_state,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        );
```
Comment (mattsse) [inline crates/net/network/src/network.rs:281]: I really don't like that we don't handle the result, 
I would prefer to update the `initial_sync_done`

if prev_state was false and future_state is true, this would be easier to read

File: crates/net/network-api/src/lib.rs [hunk 8]
```diff
@@ -49,6 +49,9 @@ pub trait NetworkInfo: Send + Sync {
 
     /// Returns `true` if the network is undergoing sync.
     fn is_syncing(&self) -> bool;
+
+    /// Returns `true` when a fresh node being bootstrapped is undergoing a Pipeline sync
```
Comment (mattsse) [inline crates/net/network-api/src/lib.rs:53]:
```suggestion
    /// Returns `true` when the node is undergoing the very first Pipeline sync.
```

File: crates/net/network/src/transactions.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/network/src/transactions.rs:857]: what do we want here?

not following

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/network/src/network.rs (hunk 7 above)
Comment (0xprames) [reply crates/net/network/src/network.rs]: sure, done

File: crates/net/network/src/network.rs [hunk 9]
```diff
@@ -247,18 +248,32 @@ impl NetworkInfo for NetworkHandle {
...
 impl SyncStateProvider for NetworkHandle {
     fn is_syncing(&self) -> bool {
         self.inner.is_syncing.load(Ordering::Relaxed)
     }
+    // used to guard the txpool
+    fn is_initially_syncing(&self) -> bool {
+        let init_sync_done = self.inner.initial_sync_done.load(Ordering::Relaxed);
+        let is_currently_syncing = self.inner.is_syncing.load(Ordering::Relaxed);
+        is_currently_syncing && !init_sync_done
+    }
```
Comment (mattsse) [inline crates/net/network/src/network.rs]: I flattened the checks here, because we only need to load `is_syncing` if initial_sync_done is false

File: crates/net/network/src/transactions.rs [hunk 10]
```diff
@@ -580,10 +579,11 @@ where
...
                 Err(err) => {
-                    // if we're syncing and the transaction is bad we ignore it, otherwise we
-                    // penalize the peer that sent the bad transaction with the assumption that the
-                    // peer should have known that this transaction is bad. (e.g. consensus rules)
-                    if err.is_bad_transaction() && !this.network.is_syncing() {
+                    // if we're initially syncing and the transaction is bad we ignore it, otherwise
+                    // we penalize the peer that sent the bad transaction with
+                    // the assumption that the peer should have known that this
+                    // transaction is bad. (e.g. consensus rules)
+                    if err.is_bad_transaction() && !this.network.is_initially_syncing() {
```
Comment (mattsse) [inline crates/net/network/src/transactions.rs]: I changed this to currently syncing

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome,

I made two changes

also the failing tests `test_tx_broadcasts_through_two_syncs` can be fixed by mocking a session established, so there's a new peer in the set.

this is the last thing to do

File: crates/net/network/src/network.rs (hunk 9 above)
Comment (0xprames) [reply crates/net/network/src/network.rs]: awesome makes sense, good catch

File: crates/net/network/src/transactions.rs (hunk 10 above)
Comment (0xprames) [reply crates/net/network/src/transactions.rs]: yeah I think we need this for every sync, good catch thanks

**Review by mattsse — APPROVED** [review]
